File: test/functional/IRDec/Rust/simple.rs
//
// RUN: rustc --crate-type=lib --emit=llvm-ir -g %s -o %t.ll
// RUN: gluc %t.ll -print-interface | FileCheck -v %s
//

#![no_main]

// CHECK: @no_mangling public func get_pizza_count() -> UInt64;
#[no_mangle] // needed when using autoimport
pub fn get_pizza_count() -> u64
{
    42
}

// CHECK: struct [[STRUCT_NAME:.*]] {
// CHECK-NEXT:    data_ptr: *UInt8,
// CHECK-NEXT:    length: UInt64,
// CHECK-NEXT: }

// CHECK: @no_mangling public func get_cc() -> [[STRUCT_NAME]];
#[no_mangle] // needed when using autoimport
pub fn get_cc() -> &'static str
{
    "Rust"
}

File: test/functional/IRDec/CPP/simple.cpp
//
// RUN: clang++ -g -c -emit-llvm %s -o %t.bc
// RUN: gluc %t.bc -print-interface | FileCheck -v %s
// RUN: gluc %s -print-interface | FileCheck -v %s
//

#include <cstddef>

class PizzaManager {
public:
    size_t getPizzaCount() const;
    size_t getPizzaMultiplier() const;
};

// CHECK: @linkage_name("{{.*}}")
// CHECK-SAME: public func getPizzaCount(this: *PizzaManager) -> UInt64;
size_t PizzaManager::getPizzaCount() const
{
    return 42 * getPizzaMultiplier();
}

// CHECK: @linkage_name("{{.*}}") public func getPizzaMultiplier(this:
// *PizzaManager) -> UInt64;
[[gnu::weak]] size_t PizzaManager::getPizzaMultiplier() const
{
    return 1;
}

// CHECK: @linkage_name("{{.*}}") public func getCC() -> *Char;
char const *getCC()
{
    return "Clang++";
}
